module/data/fields.mjs

```diff
@@ -91,3 +91,81 @@ export class IdentifierField extends foundry.data.fields.StringField {
     }
   }
 }
+
+/* -------------------------------------------- */
+
+/**
+ * A subclass of ObjectField that represents a mapping of keys to the provided DataField type.
+ *
+ * @param {DataField} type                 The class of DataField which should be embedded in this field.
+ * @param {DataFieldOptions} [options={}]  Options which configure the behavior of the field.
+ */
+export class MappingField extends foundry.data.fields.ObjectField {
+  constructor(model, options) {
+    if ( !(model instanceof foundry.data.fields.DataField) ) {
+      throw new Error("MappingField must have a DataField as its contained element");
+    }
+    super(options);
+
+    /**
+     * The embedded DataField definition which is contained in this field.
+     * @type {DataField}
+     */
+    this.model = model;
+  }
+
+  /* -------------------------------------------- */
+
+  /** @inheritdoc */
+  _cleanType(value, options) {
+    Object.values(value).forEach(v => this.model.clean(v, options));
+    return value;
+  }
+
+  /* -------------------------------------------- */
+
+  /** @inheritdoc */
+  getInitialValue(data) {
+    let keys = this.options.initialKeys;
```

Comment:
We typically try to document these extra options in `_defaults`, even if it's just declared as `initialKeys: undefined`.